src/language-js/utils.js

```diff
@@ -734,21 +734,53 @@ function returnArgumentHasLeadingComment(options, argument) {
   return false;
 }
 
-function isStringPropSafeToCoerceToIdentifier(node, options) {
+// Note: Quoting/unquoting numbers in TypeScript is not safe.
+//
+// let a = { 1: 1, 2: 2 }
+// let b = { '1': 1, '2': 2 }
+//
+// declare let aa: keyof typeof a;
+// declare let bb: keyof typeof b;
+//
+// aa = bb;
+// ^^
+// Type '"1" | "2"' is not assignable to type '1 | 2'.
+//   Type '"1"' is not assignable to type '1 | 2'.(2322)
+//
+// And in Flow, you get:
+//
+// const x = {
+//   0: 1
+//   ^ Non-string literal property keys not supported. [unsupported-syntax]
+// }
+function isStringPropSafeToUnquote(node, options) {
   return (
+    options.parser !== "json" &&
     isStringLiteral(node.key) &&
-    isIdentifierName(node.key.value) &&
     rawText(node.key).slice(1, -1) === node.key.value &&
-    options.parser !== "json" &&
-    // With `--strictPropertyInitialization`, TS treats properties with quoted names differently than unquoted ones.
-    // See https://github.com/microsoft/TypeScript/pull/20075
-    !(
-      (options.parser === "typescript" || options.parser === "babel-ts") &&
-      node.type === "ClassProperty"
-    )
+    ((isIdentifierName(node.key.value) &&
+      // With `--strictPropertyInitialization`, TS treats properties with quoted names differently than unquoted ones.
+      // See https://github.com/microsoft/TypeScript/pull/20075
+      !(
+        (options.parser === "typescript" || options.parser === "babel-ts") &&
+        node.type === "ClassProperty"
+      )) ||
+      (isSimpleNumber(node.key.value) &&
+        String(Number(node.key.value)) === node.key.value &&
+        !(
+          options.parser === "flow" ||
+          options.parser === "babel-flow" ||
+          options.parser === "typescript" ||
+          options.parser === "babel-ts"
+        )))
```

Comment:
Let's explicitly enable it only for `babel`.
Angular can't parse `{ 2: false }` in expressions.
https://stackblitz.com/edit/angular-ivy-qmpjvw?file=src%2Fapp%2Fhello.component.ts

Reply:
Vue supports it though. So alternatively, we need to disable it for `options.parser.startsWith("__ng_")`.